Custom mapping types

MartinMystikJonas opened the pull request:
See https://github.com/contributte/apitte/issues/157

MartinMystikJonas commented:
There are some CS and PHPStan errors but it is code from master so I left it unfixed so it wont mix with relevant changes

f3l1x commented on src/Core/Schema/Validation/RequestParameterValidation.php:
```diff
@@ -11,6 +11,13 @@
 
+	private array $allowedTypes = EndpointParameter::TYPES;
+
+	public function __construct(array $allowedTypes = [])
```
I'd put default value here -> EndpointParameter::TYPES.

MartinMystikJonas replied:
Ok

f3l1x commented on src/Core/Mapping/RequestParameterMapping.php:
```diff
@@ -217,7 +217,7 @@ protected function normalize($value, EndpointParameter $parameter, ITypeMapper $
 	}
 
-	protected function getMapper(string $type): ?ITypeMapper
+	public function getMapper(string $type): ?ITypeMapper
```
If it's only for tests, use reflection.

Take a look at Liberator class (https://github.com/contributte/redis/blob/master/tests/cases/DI/RedisExtension.phpt#L136-L137).

MartinMystikJonas replied:
It is for tests but it make sense for me to have this public.

f3l1x reviewed (commented):
There should be some documentation update as well.

MartinMystikJonas commented:
I will add docs

MartinMystikJonas commented:
Updated. I added docs and also exception handling for custom type validation fails. I would keep RequestParameterMapping::getMapper public - it makes sense for it to be public.

f3l1x commented:
Awesome, thanks.
